scaletest/trafficgen/run.go

```diff
@@ -0,0 +1,274 @@
+	)
+
+	// Set a deadline for stopping the text.
+	start := time.Now()
+	deadlineCtx, cancel := context.WithDeadline(ctx, start.Add(r.cfg.Duration))
+	defer cancel()
+	logger.Debug(ctx, "connect to workspace agent", slog.F("agent_id", agentID))
+
+	conn, err := r.client.WorkspaceAgentReconnectingPTY(ctx, codersdk.WorkspaceAgentReconnectingPTYOpts{
+		AgentID:   agentID,
+		Reconnect: reconnect,
+		Height:    height,
+		Width:     width,
+		Command:   "/bin/sh",
+	})
+	if err != nil {
+		logger.Error(ctx, "connect to workspace agent", slog.F("agent_id", agentID), slog.Error(err))
+		return xerrors.Errorf("connect to workspace: %w", err)
+	}
+
+	defer func() {
+		logger.Debug(ctx, "close agent connection", slog.F("agent_id", agentID))
+		_ = conn.Close()
+	}()
+
+	// Wrap the conn in a countReadWriter so we can monitor bytes sent/rcvd.
+	crw := countReadWriter{ReadWriter: conn, ctx: deadlineCtx}
+
+	// Create a ticker for sending data to the PTY.
+	tick := time.NewTicker(time.Duration(tickInterval))
+	defer tick.Stop()
+
+	// Now we begin writing random data to the pty.
+	rch := make(chan error)
+	wch := make(chan error)
+
+	go func() {
+		<-deadlineCtx.Done()
+		logger.Debug(ctx, "context deadline reached", slog.F("duration", time.Since(start)))
+	}()
+
+	// Read forever in the background.
+	go func() {
+		logger.Debug(ctx, "reading from agent", slog.F("agent_id", agentID))
+		rch <- drainContext(deadlineCtx, &crw, bytesPerTick*2)
+		logger.Debug(ctx, "done reading from agent", slog.F("agent_id", agentID))
+		conn.Close()
+		close(rch)
+	}()
+
+	// Write random data to the PTY every tick.
+	go func() {
+		logger.Debug(ctx, "writing to agent", slog.F("agent_id", agentID))
+		wch <- writeRandomData(deadlineCtx, &crw, bytesPerTick, tick.C)
+		logger.Debug(ctx, "done writing to agent", slog.F("agent_id", agentID))
+		close(wch)
+	}()
+
+	// Write until the context is canceled.
+	if wErr := <-wch; wErr != nil {
+		return xerrors.Errorf("write to pty: %w", wErr)
+	}
+	if rErr := <-rch; rErr != nil {
+		return xerrors.Errorf("read from pty: %w", rErr)
+	}
+
+	duration := time.Since(start)
+
+	logger.Info(ctx, "results",
+		slog.F("duration", duration),
+		slog.F("sent", crw.BytesWritten()),
+		slog.F("rcvd", crw.BytesRead()),
+	)
+
+	return nil
+}
+
+// Cleanup does nothing, successfully.
+func (*Runner) Cleanup(context.Context, string) error {
+	return nil
+}
+
+// drainContext drains from src until it returns io.EOF or ctx times out.
+func drainContext(ctx context.Context, src io.Reader, bufSize int64) error {
+	errCh := make(chan error, 1)
+	done := make(chan struct{})
+	go func() {
+		tmp := make([]byte, bufSize)
+		buf := bytes.NewBuffer(tmp)
+		for {
+			select {
+			case <-done:
+				return
+			default:
+				_, err := io.CopyN(buf, src, 1)
+				if ctx.Err() != nil {
+					return // context canceled while we were copying.
+				}
+				if err != nil {
+					errCh <- err
+					close(errCh)
+					return
+				}
+			}
+		}
+	}()
```

Comment:
This _might_ be racy, but I haven't been able to get goleak to complain locally.